### ISSUE TITLE
Hamster::Set#sort_by error when using a symbol

### ISSUE THREAD

**Issue by rcabralc** [hamstergem/hamster#179]
The following code works just fine:

```ruby
Hamster::Set[Object, BasicObject].sort_by { |o| o.name }
#=> Hamster::SortedSet[BasicObject, Object]
```

However, casting a block out from a symbol gives an error:

```ruby
Hamster::Set[Object, BasicObject].sort_by(&:name)
# ArgumentError: wrong number of arguments (1 for 0)
```

Giving an arity ```1``` block works (although the performance may not be optimal for expensive keys because there's no cache). But ```Symbol#to_proc```, surprisingly, returns a ```-1``` arity block. The almost equivalent code using an explicit block, similar to what ```Symbol#to_proc``` does, would be:

```ruby
Hamster::Set[Object, BasicObject].sort_by { |*a| a.shift.name(*a) }
# ArgumentError: wrong number of arguments (1 for 0)
```

For any block whose arity is not ```1```, ```SortedSet``` will just forward the block to ```Array#sort``` which expects a 2-parameters block, no matter if that code is reached from a call to ```#sort_by```. In the example above, ```#name``` ends up called with some element of the set, hence the error.

It's certainly very odd. The docs of ```Hamster::Set#sort_by``` make one think that using a symbol would work just fine. Also, the way the docs are written lead to the conclusion that the keys are retrieved once, and then the set is sorted (which is not the case). So, not sure if this is a bug in the code or in the docs.

This happens in 1.0.0 and in current master (73556ed3350).

**Comment by dubek** [comment]
Good catch!

Indeed modifying the arity condition in https://github.com/hamstergem/hamster/blob/master/lib/hamster/sorted_set.rb#L103 to:

```
if block.arity <= 1
```

fixes the problem (both with the `&:name` syntax and the weird `|*a|` block).

I'm not sure about the promises about the number of times the block is called.

**Comment by alexdowad** [comment]
Thank you for reporting this issue. A fix for the problem with
Symbol#to_proc has been pushed to master (along with a regression test).

You are very right about the block being called more times than expected.
What do you think is the most useful behavior here? Does it matter to you
that the block should only be called once? Or would just adjusting the
documentation be enough?

**Comment by rcabralc** [comment]
I think it would be desirable that the behavior followed ```Array#sort_by```. Less surprising, and a better option for expensive key computations, since currently there's no difference in using #sort or #sort_by for such keys. Then, the docs are fine.

**Comment by alexdowad** [comment]
OK, sounds good. I'll work on a patch.

**Comment by alexdowad** [comment]
Fixed in 50eaf97. @rcabralc, please check whether the fix works for you and re-open this issue if not.
